### PR TITLE
GODRIVER-2416 Completely remove the 'x/bsonx' package.

### PR DESCRIPTION
**Pull request by matthewdale**
[GODRIVER-2416](https://jira.mongodb.org/browse/GODRIVER-2416)

## Summary
Completely remove the `x/bsonx` package.

## Background & Motivation
One of the main goals of the Go Driver 2.0 changes is to reduce code duplication and restructure packages in the Go Driver to minimize import cycles. Some of that restructuring exposes unknown dependencies between packages that must be "untangled" to allow them to operate independently. The `x/bsonx` package imports some `bson` packages, complicating the restructuring of those packages. Because it is unsupported and effectively unused, we should remove it.

Completely remove the `x/bsonx` package to remove an unsupported API and simplify future restructuring of the `bson` packages. Note that this is a continuation of the work started in [GODRIVER-1953](https://jira.mongodb.org/browse/GODRIVER-1953) and [GODRIVER-1969](https://jira.mongodb.org/browse/GODRIVER-1969).

### CODE REVIEW

**Review by qingyang-hu — APPROVED** [review]
LGTM

File: benchmark/single.go [hunk 1]
```diff
@@ -93,23 +93,27 @@ func SingleFindOneByID(ctx context.Context, tm TimerManager, iters int) error {
 		return err
 	}
 	coll := db.Collection("corpus")
+
+	ids := make([]interface{}, iters)
 	for i := 0; i < iters; i++ {
-		id := int32(i)
-		res, err := coll.InsertOne(ctx, doc.Set("_id", bsonx.Int32(id)))
+		res, err := coll.InsertOne(ctx, doc)
```
Comment (kevinAlbs) [inline benchmark/single.go:99]: Suggest reverting to using the `Int32` ID. This benchmark is defined https://github.com/mongodb/specifications/blob/master/source/benchmarking/benchmarking.rst#find-one-by-id

> using sequential _id values. (1 to 10,000)

Having the driver generate an ObjectID `_id` may slightly alter the results.

**Review by kevinAlbs — APPROVED** [review]
Hooray for removed code. LGTM with a change to benchmark/single.go.

File: benchmark/single.go (hunk 1 above)
Comment (matthewdale) [reply benchmark/single.go:99]: Good catch! Will update.

**Comment by fabszabo** [comment]
You have released this breaking change in a "feature" release?! 1.11.7 -> 1.12.0?!

Even your Jira ticket says it's supposed to be in version 2!

**Comment by matthewdale** [comment]
@fabszabo Apologies for any negative impact removing the `x/bsonx` package had on you. 

The packages in the `x/` directory are [documented](https://github.com/mongodb/mongo-go-driver/blob/v1.11.7/x/README.md) as unstable and subject to breaking changes without notice. I realize that README may be easy to miss, so we're planning to make that warning more obvious in a future release (see [GODRIVER-3180](https://jira.mongodb.org/browse/GODRIVER-3180)).

Please use the following replacements from the [bson](https://pkg.go.dev/go.mongodb.org/mongo-driver/bson) package instead:
* `bsonx.Doc` ->
  * For Go representation, use `bson.D`
  * For binary representation, use `bson.Raw`
* `bsonx.Val` -> `bson.RawValue`
* `bsonx.ReadDoc` ->
  * For Go representation, use `bson.Unmarshal` with a `bson.D` as the `val` argument.
  * For binary representation, use `bson.ReadDocument`
* `bsonx.MDoc` ->
  * For Go representation, use `bson.M`
  * For binary representation, use `bson.Raw`
* `bsonx.ReadMDoc` ->
  * For Go representation, use `bson.Unmarshal` with a `bson.M` as the `val` argument.
  * For binary representation, use `bson.ReadDocument`

If you need more help replacing usages of the `x/bsonx` package, please [create a Jira ticket](https://jira.mongodb.org/secure/CreateIssue!default.jspa) in the [GODRIVER](https://jira.mongodb.org/browse/GODRIVER) project.

**Pull request merged**